src/engine/renderer/tr_surface.cpp

```diff
@@ -1053,145 +1053,159 @@ Tess_SurfaceMD5
+	transform_t *lastBone = bone + model->numBones;
 
-	if ( tess.skipTangentSpaces )
+	// Convert bones back to matrices.
+	if ( backEnd.currentEntity->e.skeleton.type == refSkeletonType_t::SK_ABSOLUTE )
 	{
-		// convert bones back to matrices
-		for (unsigned i = 0; i < model->numBones; i++ )
-		{
-			if ( backEnd.currentEntity->e.skeleton.type == refSkeletonType_t::SK_ABSOLUTE )
-			{
-				refBone_t *bone = &backEnd.currentEntity->e.skeleton.bones[ i ];
+		refBone_t *entityBone = &backEnd.currentEntity->e.skeleton.bones[ 0 ];
+		md5Bone_t *modelBone = model->bones;
 
-				TransInitRotationQuat( model->bones[ i ].rotation, &bones[ i ] );
-				TransAddTranslation( model->bones[ i ].origin, &bones[ i ] );
-				TransInverse( &bones[ i ], &bones[ i ] );
-				TransCombine( &bones[ i ], &bone->t, &bones[ i ] );
-				TransAddScale( backEnd.currentEntity->e.skeleton.scale, &bones[ i ] );
-			}
-			else
-			{
-				TransInitRotationQuat( model->bones[ i ].rotation, &bones[i] );
-				TransAddTranslation( model->bones[ i ].origin, &bones[ i ] );
-			}
-			TransInsScale( model->internalScale, &bones[ i ] );
+		for ( ; bone < lastBone; bone++,
+			modelBone++, entityBone++ )
+		{
+			TransInitRotationQuat( modelBone->rotation, bone );
+			TransAddTranslation( modelBone->origin, bone );
+			TransInverse( bone, bone );
+			TransCombine( bone, &entityBone->t, bone );
+			TransAddScale( entityScale, bone );
+			TransInsScale( modelScale, bone );
 		}
+	}
+	else if ( tess.skipTangentSpaces )
+	{
+		md5Bone_t *modelBone = model->bones;
 
-		// deform the vertices by the lerped bones
-		numVertexes = srf->numVerts;
-
-		for ( j = 0, v = srf->verts; j < numVertexes; j++, v++ )
+		for ( ; bone < lastBone; bone++,
+			modelBone++ )
 		{
-			vec3_t tmp; vec3_t pos;
-
-			VectorClear( pos );
-			for (unsigned k = 0; k < v->numWeights; k++ ) {
-				TransformPoint( &bones[ v->boneIndexes[ k ] ],
-						v->position, tmp );
-				VectorMA( pos,
-					  v->boneWeights[ k ], tmp,
-					  pos );
-
-			}
-
-			VectorCopy(pos, tess.verts[tess.numVertexes + j].xyz);
-			tess.verts[ tess.numVertexes + j ].texCoords[ 0 ] = floatToHalf( v->texCoords[ 0 ] );
-			tess.verts[ tess.numVertexes + j ].texCoords[ 1 ] = floatToHalf( v->texCoords[ 1 ] );
+			TransInitRotationQuat( modelBone->rotation, bone );
+			TransAddTranslation( modelBone->origin, bone );
+			TransInsScale( modelScale, bone );
 		}
 	}
 	else
 	{
-		tess.attribsSet |= ATTR_QTANGENT;
+		for ( ; bone < lastBone; bone++ )
+		{
+			TransInitScale( entityScale, bone );
+			TransInsScale( modelScale, bone );
+		}
+	}
+
+	glIndex_t *tessIndex = &tess.indexes [ tess.numIndexes ];
+	srfTriangle_t *surfaceTriangle = srf->triangles;
+	srfTriangle_t *lastTriangle = surfaceTriangle + srf->numTriangles;
+	md5Vertex_t *surfaceVertex = srf->verts;
+
+	for ( ; surfaceTriangle < lastTriangle; surfaceTriangle++,
+		tessIndex += 3 )
+	{
+		VectorSet( tessIndex, tess.numVertexes, tess.numVertexes, tess.numVertexes );
```

Comment:
Right, I used that Vector thing because there is a real need for function/macro that are usable for any duet/triplet/quadruplet of data. Copypasting is really annoying. So I did copypasting for MD5 instead.

Also, I just noticed that on IQM code, we can just iterate by one because both triangles and indexes are iterables the same way, and that helped me to reach 10fps on the test machine (from 9fps).